### PR TITLE
feat: adds azure support

### PR DESCRIPTION
**Pull request by prajjwaldimri**

### CODE REVIEW

**Review by abhijit-hota — APPROVED** [review]

File: pkg/proxy/proxy.go [hunk 1]
```diff
@@ -339,14 +337,26 @@ func uploadToBlobStorage(ctx context.Context, cacheID int) (*DockerGHAUploadCach
 			return nil, fmt.Errorf("no presigned URL found")
 		}
 
-		client, err := azblob.NewClientWithNoCredential(cacheEntry.BackendReserveResponse.AzureBlob.PreSignedURL, nil)
-		if err != nil {
-			return nil, fmt.Errorf("failed to create Azure Blob client: %w", err)
-		}
+		for attempt := 0; attempt < maxRetries; attempt++ {
+			contentReader := bytes.NewReader(finalBuffer.Bytes())
+			req, err := http.NewRequestWithContext(ctx, http.MethodPut, cacheEntry.BackendReserveResponse.AzureBlob.PreSignedURL, contentReader)
```
Comment (abhijit-hota) [inline pkg/proxy/proxy.go:342]: `contentReader` is not needed. You can directly pass `*bytes.Buffer` as a body.

```suggestion
			req, err := http.NewRequestWithContext(ctx, http.MethodPut, cacheEntry.BackendReserveResponse.AzureBlob.PreSignedURL, &finalBuffer)
```

File: pkg/proxy/proxy.go [hunk 2]
```diff
@@ -339,14 +337,26 @@ func uploadToBlobStorage(ctx context.Context, cacheID int) (*DockerGHAUploadCach
...
 
-		_, err = client.UploadBuffer(ctx, cacheEntry.BackendReserveResponse.AzureBlob.ContainerName, cacheEntry.BackendReserveResponse.AzureBlob.BlobName, finalBuffer.Bytes(), nil)
-		if err != nil {
-			return nil, fmt.Errorf("failed to upload to Azure Blob: %w", err)
+			resp, err := http.DefaultClient.Do(req)
+			if err == nil && resp.StatusCode >= 200 && resp.StatusCode < 300 {
+				defer resp.Body.Close()
+				break
+			} else {
+				defer resp.Body.Close()
```
Comment (abhijit-hota) [inline pkg/proxy/proxy.go:354]: Don't defer. Call directly.

**Review by abhijit-hota — CHANGES_REQUESTED** [review]
Few changes

File: pkg/proxy/proxy.go (hunk 2 above)
Comment (prajjwaldimri) [reply pkg/proxy/proxy.go]: Changed for the one in if-loop. Kept defer for the else one to be consistent with S3 case.

File: pkg/proxy/proxy.go (hunk 1 above)
Comment (prajjwaldimri) [reply pkg/proxy/proxy.go:342]: This needs testing in S3 and Azure both. Will pick it up later

File: pkg/proxy/proxy.go (hunk 2 above)
Comment (abhijit-hota) [reply pkg/proxy/proxy.go]: ack

**Review by abhijit-hota — APPROVED** [review]
LGTM. Take up the reader refactor in a separate PR after Azure support.

**Pull request merged**